google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/it/env/Emulator.java

```diff
@@ -0,0 +1,214 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.it.env;
+
+import com.google.api.gax.core.NoCredentialsProvider;
+import com.google.api.gax.grpc.GrpcTransportChannel;
+import com.google.api.gax.rpc.ClientSettings;
+import com.google.api.gax.rpc.FixedTransportChannelProvider;
+import com.google.bigtable.admin.v2.InstanceName;
+import com.google.cloud.bigtable.admin.v2.BigtableTableAdminClient;
+import com.google.cloud.bigtable.admin.v2.BigtableTableAdminSettings;
+import com.google.cloud.bigtable.data.v2.BigtableDataClient;
+import com.google.cloud.bigtable.data.v2.BigtableDataSettings;
+import com.google.common.io.CharStreams;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** Java wrapper around the gcloud bigtable emulator. */
+// TODO(igorbernstein): Clean this up and externalize this in a separate artifact
+// TODO(igorbernstein): Stop depending on gcloud for the binary, instead wrap it in a jar.
+class Emulator {
+  private static final Logger LOGGER = Logger.getLogger(Emulator.class.getName());
+
+  private final Path executable;
+  private Process process;
+  private ManagedChannel channel;
+  private BigtableTableAdminClient tableAdminClient;
+  private BigtableDataClient dataClient;
+
+  private static final InstanceName INSTANCE_NAME =
+      InstanceName.of("fake-project", "fake-instance");
+
+  // Use the gcloud installed emulator
+  static Emulator createGCloud() {
+    final Path gcloudSdkPath;
+
+    try {
+      gcloudSdkPath = getGcloudSdkPath();
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to get the gcloud SDK path. Is it installed?", e);
+    }
+
+    Path emulatorPath =
+        gcloudSdkPath.resolve(Paths.get("platform", "bigtable-emulator", "cbtemulator"));
+
+    if (!Files.exists(emulatorPath)) {
+      throw new RuntimeException(
+          "cbtemulator is not installed, please install with `gcloud components install bigtable`");
+    }
+
+    return new Emulator(emulatorPath);
+  }
+
+  private Emulator(Path executable) {
+    this.executable = executable;
+  }
+
+  void start() throws Exception {
+    int availablePort = getAvailablePort();
+
+    process = Runtime.getRuntime().exec(executable + " -port " + "" + availablePort);
```

Comment:
@pongad isn't one of your main concerns about emulators the fact that doing process stuff in Java is problematic?

Reply:
This implementation should address most of the concerns:
- it is executing the binary directly so java can still kill the process
- all output is copied to the logger
- nothing is installed behind the user's back